lib/semantics/symbol.h

```diff
@@ -582,8 +582,23 @@ class Symbol {
         details_);
   }
 
-  bool operator==(const Symbol &that) const { return this == &that; }
-  bool operator!=(const Symbol &that) const { return this != &that; }
+  // For the purposes of comparing type parameter expressions while
+  // testing the compatibility of procedure characteristics, two
+  // object dummy arguments with the same name are considered equal.
+  bool operator==(const Symbol &that) const {
+    if (this == &that) {
+      return true;
+    } else if (name() != that.name()) {
+      return false;
+    } else if (const auto *object{detailsIf<ObjectEntityDetails>()}) {
+      if (const auto *thatObject{that.detailsIf<ObjectEntityDetails>()}) {
+        return object->isDummy() && thatObject->isDummy();
+      }
+    }
+    return false;
+  }
```

Comment:
This definition of `operator==` is not consistent with `operator<`. So you could have two symbols, `x` and `y,` such that `x < y` and `x == y`. It probably wouldn't work to put them in the same `std::set` or `std::map`.

If think it would be better to add a new function for this version of equivalence.

Reply:
Ok, fine.  I'm going to have to replace all the names of the `operator==` instances in expressions that end up here, too.

Reply:
Done in branch pmk-call12, work in progress.